File: python-client/p2pshare/network/peer.py
"""Peer connection state machine.

Manages the lifecycle of a single peer-to-peer TCP connection:
  1. TCP connect
  2. Handshake (HELLO/HELLO_ACK/AUTH/AUTH_ACK)
  3. Encrypted message exchange
  4. Graceful close
"""

import socket
import json
import math
import threading

from p2pshare.crypto.keys import (
    b64url_encode,
    b64url_decode,
    bytes_to_public_key,
    public_key_to_bytes,
    peer_id_from_public_key,
    fingerprint,
)
from p2pshare.crypto.handshake import HandshakeInitiator, HandshakeResponder
from p2pshare.crypto.session import SessionCipher
from p2pshare.network.transport import (
    send_json,
    recv_json,
    send_encrypted,
    recv_encrypted,
)
from p2pshare.protocol import messages
from p2pshare.protocol.files import (
    handle_list_request,
    build_file_chunks,
    reassemble_file,
    verify_received_file,
)
from p2pshare.protocol.migration import verify_migration
from p2pshare.storage.vault import Vault
from p2pshare.storage.file_store import FileStore
from p2pshare.errors import (
    HandshakeError,
    AuthenticationError,
    IntegrityError,
    ConsentDeniedError,
    P2PShareError,
)


class PeerConnection:
    """A single authenticated, encrypted peer connection."""

    def __init__(self, sock: socket.socket, vault: Vault, file_store: FileStore,
                 consent_callback=None):
        self.sock = sock
        self.vault = vault
        self.file_store = file_store
        self.cipher: SessionCipher | None = None
        self.peer_id: str = ""
        self.peer_pk = None
        self.seq = 0
        self._consent_callback = consent_callback or self._default_consent

    def _next_seq(self) -> int:
        self.seq += 1
        return self.seq

    def connect_as_initiator(self, host: str, port: int) -> None:
        """Initiate a connection and perform the handshake."""
        self.sock.connect((host, port))
        hs = HandshakeInitiator(self.vault.identity_sk, self.vault.identity_pk)

        # Step 1: send HELLO
        hello = hs.create_hello()
        send_json(self.sock, hello)

        # Step 2: receive HELLO_ACK
        hello_ack = recv_json(self.sock)
        hs.process_hello_ack(hello_ack)

        # Derive session key
        session_key = hs.derive_session()

        # Step 3: send AUTH
        auth, self.cipher = hs.create_auth(session_key)
        send_encrypted(self.sock, self.cipher, auth)

        # Step 4: receive AUTH_ACK
        auth_ack = recv_encrypted(self.sock, self.cipher)
        hs.verify_auth_ack(auth_ack)

        self.peer_id = b64url_encode(public_key_to_bytes(hs.peer_identity_pk))
        self.peer_pk = hs.peer_identity_pk
        self._handle_new_peer()

    def accept_as_responder(self) -> None:
        """Accept an incoming connection and perform the handshake."""
        hs = HandshakeResponder(self.vault.identity_sk, self.vault.identity_pk)

        # Step 1: receive HELLO
        hello = recv_json(self.sock)
        hs.process_hello(hello)

        # Step 2: send HELLO_ACK
        hello_ack = hs.create_hello_ack()
        send_json(self.sock, hello_ack)

        # Derive session key
        session_key = hs.derive_session()

        # Step 3: receive AUTH
        auth = recv_encrypted(self.sock, SessionCipher(session_key))
        hs.verify_auth(auth)

        # Step 4: send AUTH_ACK
        auth_ack, self.cipher = hs.create_auth_ack(session_key)
        send_encrypted(self.sock, self.cipher, auth_ack)

        self.peer_id = b64url_encode(public_key_to_bytes(hs.peer_identity_pk))
        self.peer_pk = hs.peer_identity_pk
        self._handle_new_peer()

    def _handle_new_peer(self) -> None:
        """Handle a newly authenticated peer (TOFU contact add)."""
        contact = self.vault.get_contact(self.peer_id)
        if contact is None:
            pk_bytes = public_key_to_bytes(self.peer_pk)
            self.vault.add_contact(self.peer_id, pk_bytes)
            self.vault.save()
            fp = fingerprint(self.peer_pk)
            print(f"[+] New peer added to contacts: {self.peer_id[:8]}... (fingerprint: {fp})")

    def send_msg(self, msg: dict) -> None:
        """Send an encrypted message."""
        send_encrypted(self.sock, self.cipher, msg)

    def recv_msg(self) -> dict:
        """Receive and decrypt a message."""
        return recv_encrypted(self.sock, self.cipher)

    def request_file_list(self) -> list[dict]:
        """Request and return the peer's file list."""
        req = messages.list_request(self._next_seq())
        self.send_msg(req)
        resp = self.recv_msg()
        if resp.get("type") == "ERROR":
            print(f"[!] Error: {resp.get('message', 'Unknown error')}")
            return []
        return resp.get("files", [])

    def request_file(self, file_id: str) -> bytes | None:
        """Request a file from the peer. Returns file data or None if declined."""
        req = messages.transfer_request(self._next_seq(), file_id, self.vault.peer_id)
        self.send_msg(req)

        resp = self.recv_msg()
        if resp.get("type") == "TRANSFER_RESPONSE":
            if not resp.get("accepted"):
                print(f"[!] Transfer declined by peer: {resp.get('reason', '')}")
                return None
        elif resp.get("type") == "ERROR":
            print(f"[!] Error: {resp.get('message', 'Unknown error')}")
            return None

        # Receive chunks
        chunks = []
        while True:
            msg = self.recv_msg()
            if msg.get("type") == "TRANSFER_CHUNK":
                chunks.append(msg)
            elif msg.get("type") == "TRANSFER_COMPLETE":
                break
            elif msg.get("type") == "ERROR":
                print(f"[!] Error during transfer: {msg.get('message')}")
                return None

        return reassemble_file(chunks)

    def send_file(self, file_id: str) -> None:
        """Send a file to the peer (called after consent is granted)."""
        chunk_msgs = build_file_chunks(self.file_store, file_id, self.seq)
        for msg in chunk_msgs:
            self.send_msg(msg)
            self.seq = msg["seq"]

    def handle_incoming(self) -> None:
        """Main loop for handling incoming messages from a connected peer."""
        seq_counter = [self.seq]
        try:
            while True:
                msg = self.recv_msg()
                msg_type = msg.get("type")

                if msg_type == "LIST_REQUEST":
                    resp = handle_list_request(msg, self.file_store, seq_counter)
                    self.send_msg(resp)

                elif msg_type == "TRANSFER_REQUEST":
                    file_id = msg.get("file_id", "")
                    requester = msg.get("requester_id", "")[:8]
                    incoming_manifest = msg.get("manifest")

                    # Resolve file_id prefix to full file_id
                    if not self.file_store.has_file(file_id):
                        for m in self.file_store.list_manifests():
                            if m["file_id"].startswith(file_id):
                                file_id = m["file_id"]
                                break

                    if self.file_store.has_file(file_id):
                        # Pull flow: requester wants to download this file from us
                        manifest = self.file_store.load_manifest(file_id)
                        filename = manifest.get("filename", file_id[:8])
                        accepted = self._consent_callback(
                            f"Peer {requester}... is requesting '{filename}'. Send it? [y/n]: "
                        )
                        seq_counter[0] += 1
                        if accepted:
                            self.send_msg(messages.transfer_response(seq_counter[0], file_id, True))
                            self.seq = seq_counter[0]
                            self.send_file(file_id)
                        else:
                            self.send_msg(messages.transfer_response(seq_counter[0], file_id, False, "User declined"))

                    elif incoming_manifest is not None:
                        # Push flow: sender wants to push a file to us
                        filename = incoming_manifest.get("filename", file_id[:8])
                        size_bytes = incoming_manifest.get("size_bytes", 0)
                        accepted = self._consent_callback(
                            f"Peer {requester}... wants to send you '{filename}' ({size_bytes} bytes). Accept? [y/n]: "
                        )
                        seq_counter[0] += 1
                        if not accepted:
                            self.send_msg(messages.transfer_response(seq_counter[0], file_id, False, "User declined"))
                        else:
                            self.send_msg(messages.transfer_response(seq_counter[0], file_id, True))
                            self.seq = seq_counter[0]
                            # Receive chunks from sender
                            chunks = []
                            complete_msg = None
                            while True:
                                chunk_msg = self.recv_msg()
                                if chunk_msg.get("type") == "TRANSFER_CHUNK":
                                    chunks.append(chunk_msg)
                                elif chunk_msg.get("type") == "TRANSFER_COMPLETE":
                                    complete_msg = chunk_msg
                                    break
                                elif chunk_msg.get("type") == "ERROR":
                                    print(f"[!] Error during receive: {chunk_msg.get('message')}")
                                    break
                            if complete_msg is not None and chunks:
                                file_data = reassemble_file(chunks)
                                owner_id = incoming_manifest.get("owner_id", "")
                                contact = self.vault.get_contact(owner_id)
                                if contact:
                                    owner_pk_bytes = b64url_decode(contact["public_key"])
                                    try:
                                        verify_received_file(file_data, incoming_manifest, owner_pk_bytes)
                                        print(f"[+] File integrity and owner signature verified.")
                                    except Exception as e:
                                        print(f"[!] {e}")
                                        continue
                                else:
                                    print(f"[!] Warning: Cannot verify owner — {owner_id[:8]}... not in contacts.")
                                self.file_store.store_file(file_id, file_data, incoming_manifest)
                                print(f"[+] Received file '{filename}' saved.")

                    else:
                        self.send_msg(messages.error_msg(
                            self._next_seq(), messages.FILE_NOT_FOUND,
                            f"File '{file_id[:8]}...' is not available."))

                elif msg_type == "KEY_MIGRATION":
                    self._handle_migration(msg, seq_counter)

                elif msg_type == "ERROR":
                    print(f"[!] Peer error: [{msg.get('code')}] {msg.get('message')}")

                else:
                    print(f"[?] Unknown message type: {msg_type}")

        except P2PShareError as e:
            print(f"[!] Connection error: {e}")
        except Exception:
            pass  # Connection closed

    def _handle_migration(self, msg: dict, seq_counter: list) -> None:
        """Process a KEY_MIGRATION announcement."""
        old_peer_id = msg.get("old_peer_id", "")
        contact = self.vault.get_contact(old_peer_id)
        if contact is None:
            seq_counter[0] += 1
            self.send_msg(messages.migration_ack(seq_counter[0], False, "Unknown peer"))
            return

        old_pk_bytes = b64url_decode(contact["public_key"])
        try:
            new_pk_bytes = verify_migration(msg, old_pk_bytes)
        except Exception as e:
            print(f"[!] {e}")
            seq_counter[0] += 1
            self.send_msg(messages.migration_ack(seq_counter[0], False, str(e)))
            return

        new_peer_id = msg.get("new_peer_id", "")
        self.vault.update_contact_key(old_peer_id, new_peer_id, new_pk_bytes)
        self.vault.save()
        print(f"[+] Key migration accepted for peer {old_peer_id[:8]}... → {new_peer_id[:8]}...")
        seq_counter[0] += 1
        self.send_msg(messages.migration_ack(seq_counter[0], True))

    @staticmethod
    def _default_consent(prompt: str) -> bool:
        """Default consent callback: blocking stdin prompt."""
        try:
            answer = input(prompt).strip().lower()
            return answer in ("y", "yes")
        except EOFError:
            return False

    def close(self) -> None:
        """Close the connection."""
        try:
            self.sock.close()
        except Exception:
            pass

File: go-client/internal/network/peer.go
package network

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"strings"

	"github.com/cisc468/p2p-project/internal/crypto"
	"github.com/cisc468/p2p-project/internal/protocol"
	"github.com/cisc468/p2p-project/internal/storage"
)

// ConsentFunc is a callback that prompts the user for consent. Returns true if accepted.
type ConsentFunc func(prompt string) bool

// PeerConnection represents a single authenticated, encrypted peer connection.
type PeerConnection struct {
	Conn      net.Conn
	Vault     *storage.Vault
	FileStore *storage.FileStore
	Cipher    *crypto.SessionCipher
	PeerID    string
	PeerPK    []byte // raw 32-byte Ed25519 public key
	Seq       int
	Consent   ConsentFunc
}

// NewPeerConnection creates a new peer connection wrapper.
func NewPeerConnection(conn net.Conn, vault *storage.Vault, fs *storage.FileStore, consent ConsentFunc) *PeerConnection {
	if consent == nil {
		consent = DefaultConsent
	}
	return &PeerConnection{
		Conn:      conn,
		Vault:     vault,
		FileStore: fs,
		Consent:   consent,
	}
}

func (pc *PeerConnection) nextSeq() int {
	pc.Seq++
	return pc.Seq
}

// ConnectAsInitiator performs the 4-step handshake as the initiator.
func (pc *PeerConnection) ConnectAsInitiator() error {
	hs, err := crypto.NewHandshakeInitiator(pc.Vault.IdentitySK, pc.Vault.IdentityPK)
	if err != nil {
		return err
	}

	// Step 1: send HELLO
	hello, err := hs.CreateHello()
	if err != nil {
		return err
	}
	if err := SendJSON(pc.Conn, hello); err != nil {
		return err
	}

	// Step 2: receive HELLO_ACK
	helloAck, err := RecvJSON(pc.Conn)
	if err != nil {
		return err
	}
	if err := hs.ProcessHelloAck(helloAck); err != nil {
		return err
	}

	// Derive session key
	sessionKey, err := hs.DeriveSession()
	if err != nil {
		return err
	}

	// Step 3: send AUTH
	auth, cipher, err := hs.CreateAuth(sessionKey)
	if err != nil {
		return err
	}
	pc.Cipher = cipher
	if err := SendEncrypted(pc.Conn, pc.Cipher, auth); err != nil {
		return err
	}

	// Step 4: receive AUTH_ACK
	authAck, err := RecvEncrypted(pc.Conn, pc.Cipher)
	if err != nil {
		return err
	}
	if err := hs.VerifyAuthAck(authAck); err != nil {
		return err
	}

	pc.PeerID = crypto.PeerIDFromPublicKey(hs.PeerIdentityPK)
	pc.PeerPK = crypto.PublicKeyToBytes(hs.PeerIdentityPK)
	pc.handleNewPeer()
	return nil
}

// AcceptAsResponder performs the 4-step handshake as the responder.
func (pc *PeerConnection) AcceptAsResponder() error {
	hs, err := crypto.NewHandshakeResponder(pc.Vault.IdentitySK, pc.Vault.IdentityPK)
	if err != nil {
		return err
	}

	// Step 1: receive HELLO
	hello, err := RecvJSON(pc.Conn)
	if err != nil {
		return err
	}
	if err := hs.ProcessHello(hello); err != nil {
		return err
	}

	// Step 2: send HELLO_ACK
	helloAck, err := hs.CreateHelloAck()
	if err != nil {
		return err
	}
	if err := SendJSON(pc.Conn, helloAck); err != nil {
		return err
	}

	// Derive session key
	sessionKey, err := hs.DeriveSession()
	if err != nil {
		return err
	}

	// Step 3: receive AUTH (need a temporary cipher for the first encrypted message)
	tmpCipher, err := crypto.NewSessionCipher(sessionKey)
	if err != nil {
		return err
	}
	authMsg, err := RecvEncrypted(pc.Conn, tmpCipher)
	if err != nil {
		return err
	}
	if err := hs.VerifyAuth(authMsg); err != nil {
		return err
	}

	// Step 4: send AUTH_ACK
	authAck, cipher, err := hs.CreateAuthAck(sessionKey)
	if err != nil {
		return err
	}
	pc.Cipher = cipher
	if err := SendEncrypted(pc.Conn, pc.Cipher, authAck); err != nil {
		return err
	}

	pc.PeerID = crypto.PeerIDFromPublicKey(hs.PeerIdentityPK)
	pc.PeerPK = crypto.PublicKeyToBytes(hs.PeerIdentityPK)
	pc.handleNewPeer()
	return nil
}

func (pc *PeerConnection) handleNewPeer() {
	contact := pc.Vault.GetContact(pc.PeerID)
	if contact == nil {
		pc.Vault.AddContact(pc.PeerID, pc.PeerPK, "")
		_ = pc.Vault.Save()
		peerPK, _ := crypto.BytesToPublicKey(pc.PeerPK)
		fp := crypto.Fingerprint(peerPK)
		fmt.Printf("[+] New peer added to contacts: %s... (fingerprint: %s)\n", pc.PeerID[:8], fp)
	}
}

// SendMsg sends an encrypted message.
func (pc *PeerConnection) SendMsg(msg map[string]interface{}) error {
	return SendEncrypted(pc.Conn, pc.Cipher, msg)
}

// RecvMsg receives and decrypts a message.
func (pc *PeerConnection) RecvMsg() (map[string]interface{}, error) {
	return RecvEncrypted(pc.Conn, pc.Cipher)
}

// RequestFileList requests the peer's file list.
func (pc *PeerConnection) RequestFileList() ([]map[string]interface{}, error) {
	req := protocol.ListRequest(pc.nextSeq(), nil)
	if err := pc.SendMsg(req); err != nil {
		return nil, err
	}
	resp, err := pc.RecvMsg()
	if err != nil {
		return nil, err
	}
	if resp["type"] == "ERROR" {
		return nil, fmt.Errorf("peer error: %v", resp["message"])
	}
	filesRaw, _ := resp["files"].([]interface{})
	var files []map[string]interface{}
	for _, f := range filesRaw {
		if fm, ok := f.(map[string]interface{}); ok {
			files = append(files, fm)
		}
	}
	return files, nil
}

// RequestFile requests a file from the peer (pull/download flow).
func (pc *PeerConnection) RequestFile(fileID string) ([]byte, error) {
	req := protocol.TransferRequest(pc.nextSeq(), fileID, pc.Vault.PeerID(), nil)
	if err := pc.SendMsg(req); err != nil {
		return nil, err
	}

	resp, err := pc.RecvMsg()
	if err != nil {
		return nil, err
	}
	if resp["type"] == "TRANSFER_RESPONSE" {
		accepted, _ := resp["accepted"].(bool)
		if !accepted {
			reason, _ := resp["reason"].(string)
			return nil, fmt.Errorf("transfer declined by peer: %s", reason)
		}
	} else if resp["type"] == "ERROR" {
		return nil, fmt.Errorf("peer error: %v", resp["message"])
	}

	// Receive chunks
	var chunks []map[string]interface{}
	for {
		msg, err := pc.RecvMsg()
		if err != nil {
			return nil, err
		}
		msgType, _ := msg["type"].(string)
		if msgType == "TRANSFER_CHUNK" {
			chunks = append(chunks, msg)
		} else if msgType == "TRANSFER_COMPLETE" {
			break
		} else if msgType == "ERROR" {
			return nil, fmt.Errorf("error during transfer: %v", msg["message"])
		}
	}

	return protocol.ReassembleFile(chunks)
}

// SendFile sends a file to the peer after consent is granted.
func (pc *PeerConnection) SendFile(fileID string) error {
	chunkMsgs, err := protocol.BuildFileChunks(pc.FileStore, fileID, pc.Seq)
	if err != nil {
		return err
	}
	for _, msg := range chunkMsgs {
		if err := pc.SendMsg(msg); err != nil {
			return err
		}
		if s, ok := msg["seq"].(float64); ok {
			pc.Seq = int(s)
		}
	}
	return nil
}

// HandleIncoming is the main loop for handling incoming messages.
func (pc *PeerConnection) HandleIncoming() {
	for {
		msg, err := pc.RecvMsg()
		if err != nil {
			return // Connection closed
		}
		msgType, _ := msg["type"].(string)

		switch msgType {
		case "LIST_REQUEST":
			resp, err := protocol.HandleListRequest(msg, pc.FileStore, &pc.Seq)
			if err != nil {
				pc.SendMsg(protocol.ErrorMsg(pc.nextSeq(), protocol.ProtocolErrorCode, err.Error()))
				continue
			}
			pc.SendMsg(resp)

		case "TRANSFER_REQUEST":
			pc.handleTransferRequest(msg)

		case "KEY_MIGRATION":
			pc.handleMigration(msg)

		case "ERROR":
			code, _ := msg["code"].(string)
			message, _ := msg["message"].(string)
			fmt.Printf("[!] Peer error: [%s] %s\n", code, message)

		default:
			fmt.Printf("[?] Unknown message type: %s\n", msgType)
		}
	}
}

func (pc *PeerConnection) handleTransferRequest(msg map[string]interface{}) {
	fileID, _ := msg["file_id"].(string)
	requesterID, _ := msg["requester_id"].(string)

	// Resolve file_id prefix to full file_id
	if !pc.FileStore.HasFile(fileID) {
		if manifests, err := pc.FileStore.ListManifests(); err == nil {
			for _, m := range manifests {
				if strings.HasPrefix(m.FileID, fileID) {
					fileID = m.FileID
					break
				}
			}
		}
	}

	short := requesterID
	if len(short) > 8 {
		short = short[:8]
	}

	if pc.FileStore.HasFile(fileID) {
		// Pull flow: requester wants to download this file from us
		manifest, err := pc.FileStore.LoadManifest(fileID)
		if err != nil {
			pc.SendMsg(protocol.ErrorMsg(pc.nextSeq(), protocol.FileNotFound, err.Error()))
			return
		}
		prompt := fmt.Sprintf("Peer %s... is requesting '%s'. Send it? [y/n]: ", short, manifest.Filename)
		if pc.Consent(prompt) {
			pc.SendMsg(protocol.TransferResponse(pc.nextSeq(), fileID, true, ""))
			pc.SendFile(fileID)
		} else {
			pc.SendMsg(protocol.TransferResponse(pc.nextSeq(), fileID, false, "User declined"))
		}
	} else if incomingManifest, ok := msg["manifest"].(map[string]interface{}); ok {
		// Push flow: sender wants to push a file to us
		filename, _ := incomingManifest["filename"].(string)
		sizeF, _ := incomingManifest["size_bytes"].(float64)
		if filename == "" {
			filename = fileID
			if len(filename) > 8 {
				filename = filename[:8]
			}
		}
		prompt := fmt.Sprintf("Peer %s... wants to send you '%s' (%.0f bytes). Accept? [y/n]: ", short, filename, sizeF)
		if pc.Consent(prompt) {
			pc.SendMsg(protocol.TransferResponse(pc.nextSeq(), fileID, true, ""))
			pc.receiveFile(fileID, incomingManifest)
		} else {
			pc.SendMsg(protocol.TransferResponse(pc.nextSeq(), fileID, false, "User declined"))
		}
	} else {
		pc.SendMsg(protocol.ErrorMsg(pc.nextSeq(), protocol.FileNotFound,
			fmt.Sprintf("File '%s...' is not available.", fileID[:8])))
	}
}

func (pc *PeerConnection) receiveFile(fileID string, manifestRaw map[string]interface{}) {
	var chunks []map[string]interface{}
loop:
	for {
		msg, err := pc.RecvMsg()
		if err != nil {
			return
		}
		msgType, _ := msg["type"].(string)
		switch msgType {
		case "TRANSFER_CHUNK":
			chunks = append(chunks, msg)
		case "TRANSFER_COMPLETE":
			break loop
		case "ERROR":
			fmt.Printf("[!] Error during receive: %v\n", msg["message"])
			return
		}
	}

	if len(chunks) == 0 {
		return
	}

	fileData, err := protocol.ReassembleFile(chunks)
	if err != nil {
		fmt.Printf("[!] Failed to reassemble file: %v\n", err)
		return
	}

	data, _ := json.Marshal(manifestRaw)
	var manifest storage.Manifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		fmt.Printf("[!] Invalid manifest: %v\n", err)
		return
	}

	ownerID := manifest.OwnerID
	contact := pc.Vault.GetContact(ownerID)
	if contact != nil {
		ownerPKBytes, _ := crypto.B64URLDecode(contact.PublicKey)
		if err := protocol.VerifyReceivedFile(fileData, &manifest, ownerPKBytes); err != nil {
			fmt.Printf("[!] %v\n", err)
			return
		}
		fmt.Println("[+] File integrity and owner signature verified.")
	} else {
		ownerShort := ownerID
		if len(ownerShort) > 8 {
			ownerShort = ownerShort[:8]
		}
		fmt.Printf("[!] Warning: Cannot verify owner — %s... not in contacts.\n", ownerShort)
	}

	if err := pc.FileStore.StoreFile(fileID, fileData, &manifest); err != nil {
		fmt.Printf("[!] Failed to store file: %v\n", err)
		return
	}
	fmt.Printf("[+] Received file '%s' saved.\n", manifest.Filename)
}

func (pc *PeerConnection) handleMigration(msg map[string]interface{}) {
	oldPeerID, _ := msg["old_peer_id"].(string)
	contact := pc.Vault.GetContact(oldPeerID)
	if contact == nil {
		pc.SendMsg(protocol.MigrationAck(pc.nextSeq(), false, "Unknown peer"))
		return
	}

	oldPKBytes, err := crypto.B64URLDecode(contact.PublicKey)
	if err != nil {
		pc.SendMsg(protocol.MigrationAck(pc.nextSeq(), false, "Invalid stored key"))
		return
	}

	newPKBytes, err := protocol.VerifyMigration(msg, oldPKBytes)
	if err != nil {
		fmt.Printf("[!] %v\n", err)
		pc.SendMsg(protocol.MigrationAck(pc.nextSeq(), false, err.Error()))
		return
	}

	newPeerID, _ := msg["new_peer_id"].(string)
	pc.Vault.UpdateContactKey(oldPeerID, newPeerID, newPKBytes)
	pc.Vault.Save()
	fmt.Printf("[+] Key migration accepted for peer %s... → %s...\n", oldPeerID[:8], newPeerID[:8])
	pc.SendMsg(protocol.MigrationAck(pc.nextSeq(), true, ""))
}

// DefaultConsent prompts the user via stdin.
func DefaultConsent(prompt string) bool {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	answer, _ := reader.ReadString('\n')
	answer = strings.TrimSpace(strings.ToLower(answer))
	return answer == "y" || answer == "yes"
}

// Close closes the connection.
func (pc *PeerConnection) Close() {
	pc.Conn.Close()
}
